torch/csrc/jit/passes/constant_pooling.cpp

```diff
@@ -0,0 +1,53 @@
+#include "torch/csrc/jit/ir.h"
+#include <unordered_set>
+#include "torch/csrc/jit/interned_strings.h"
+#include "torch/csrc/jit/passes/constant_pooling.h"
+#include "torch/csrc/jit/node_hashing.h"
+
+namespace torch { namespace jit {
+
+namespace {
+
+//Very similar to the common subexpression elimination pass
+//Move all constants to the beginning of the graph, and deduplicate
+void ConstantPooling(Block * block, std::unordered_set<Node*, HashNodeCSE, EqualNodeCSE>& constants) {
+  for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+    auto node = *it;
+    // node may be moved to a different block so advance iterator now
+    ++it;
+    if (!node->blocks().empty()) {
+      // Traverse sub-blocks.
+      for (auto block : node->blocks()) {
+        ConstantPooling(block, constants);
+      }
+      continue;
+    }
+
+    if (node->kind() != prim::Constant) {
+      continue;
+    }
+
+    auto first_node = node->owningGraph()->block()->nodes().front();
```

Comment:
Shouldn't this move the constant to the first node? Not after the first node?
```
node->moveBefore(node->owningGraph()->nodes().front());
```

Reply:
Also it would be good to guarantee after this pass runs that all the constants are at a block at the beginning of the function. This would make it easy for the interpreter to internally do an optimization where it would pre-process the constants into a list of IValues and just copy them into the first C registers when the function begins to execute.